linera-core/src/worker.rs

```diff
@@ -761,85 +824,131 @@ where
-                // but we just checked that it is.
-                let timeout_certificate = certificate.into();
-                self.process_timeout(timeout_certificate).await?
+            CertificateValue::ValidatedBlock(_) => {
+                let certificate = ValidatedBlockCertificate::from(certificate);
+                self.handle_validated_certificate(certificate, blobs).await
             }
-        };
+            CertificateValue::Timeout(_) => {
+                let certificate = TimeoutCertificate::from(certificate);
+                self.handle_timeout_certificate(certificate).await
+            }
+        }
+    }
 
+    /// Processes a confirmed block certificate.
+    #[instrument(skip_all, fields(
+        nick = self.nickname,
+        chain_id = format!("{:.8}", certificate.executed_block().block.chain_id),
+        height = %certificate.executed_block().block.height,
+    ))]
+    pub async fn handle_confirmed_certificate(
+        &self,
+        certificate: ConfirmedBlockCertificate,
+        blobs: Vec<Blob>,
+        notify_when_messages_are_delivered: Option<oneshot::Sender<()>>,
+    ) -> Result<(ChainInfoResponse, NetworkActions), WorkerError> {
+        trace!("{} <-- {:?}", self.nickname, certificate);
         #[cfg(with_metrics)]
-        if !duplicated {
+        {
+            let confirmed_transactions = (certificate.executed_block().block.incoming_bundles.len()
+                + certificate.executed_block().block.operations.len())
+                as u64;
+
             NUM_ROUNDS_IN_CERTIFICATE
-                .with_label_values(&[log_str, round.type_name()])
-                .observe(round.number() as f64);
+                .with_label_values(&[
+                    certificate.inner().to_log_str(),
+                    certificate.round.type_name(),
+                ])
+                .observe(certificate.round.number() as f64);
             if confirmed_transactions > 0 {
                 TRANSACTION_COUNT
                     .with_label_values(&[])
                     .inc_by(confirmed_transactions);
             }
         }
+
+        self.process_confirmed_block(certificate, &blobs, notify_when_messages_are_delivered)
+            .await
+    }
+
+    /// Processes a validated block certificate.
+    #[instrument(skip_all, fields(
+        nick = self.nickname,
+        chain_id = format!("{:.8}", certificate.executed_block().block.chain_id),
+        height = %certificate.executed_block().block.height,
+    ))]
+    pub async fn handle_validated_certificate(
+        &self,
+        certificate: ValidatedBlockCertificate,
+        blobs: Vec<Blob>,
+    ) -> Result<(ChainInfoResponse, NetworkActions), WorkerError> {
+        trace!("{} <-- {:?}", self.nickname, certificate);
+
+        #[cfg(with_metrics)]
+        let round = certificate.round;
+        #[cfg(with_metrics)]
+        let cert_str = certificate.inner().to_log_str();
+
+        let (info, actions, _duplicated) =
```

Comment:
(Not in this PR, but: Now I'm wondering why we don't check for duplicates in the other cases.)